nexus/sdk-python.md

```diff
@@ -0,0 +1,330 @@
+
+# 2. Implementing a Nexus operation
+
+Traditionally, in order to define an RPC/HTTP handler, you specify a function to act as the handler, and you arrange for a service or operation instance with a lifetime equal to that of the server, containing state created at server-start time, to be in-scope in the function body. For Python examples, see [gRPC,](https://github.com/grpc/grpc/blob/4e9357bca1408596663a218c0c608a4c0560a867/examples/python/route_guide/route_guide_server.py#L65) [xmlrpc [stdlib]](https://docs.python.org/3/library/xmlrpc.server.html#simplexmlrpcserver-example), [starlette/FastAPI](https://www.starlette.io/applications/#storing-state-on-the-app-instance), [Django class-based-views](https://docs.djangoproject.com/en/5.1/topics/class-based-views/). So, a Python developer writing an RPC/HTTP handler will expect to write something like
+
+```python
+# This code (from Python gRPC docs) is not part of the proposal
+class MyService:  # instantiated at server-start time
+    def my_handler(self, my_input, context) -> MyResult: ...
+```
+
+However, unlike traditional RPC/HTTP handlers, a Nexus operation represents a long-running execution and must support three additional verbs (`get_info`, `get_result`, `cancel`), and must support `start` returning both synchronously and asynchronously (including making that decision at request-handling time).
+
+## 2a. Fully manual class-based operation definition
+
+The “fully manual” way we will support this in Nexus Python is that a user will define a class for the service, and a class for each operation.
+
+**[ALTERNATIVES CONSIDERED]** None. Giving the user the option to define classes at both levels for maximum implementation flexibility is the natural choice for Python (as it was in Java).
+
+```python
+import nexusrpc.handler
+
+@nexusrpc.handler.service(interface=interface.MyNexusService, name="my-service")  # import-time check that interface was implemented
+class MyNexusService: # instantiated when instantiating worker
+    # User is free to define custom constructor
+
+    @nexusrpc.handler.operation(name="my-invalid-python-identifier")
+    def echo(self) -> nexusrpc.handler.Operation[EchoInput, EchoOutput]:  # typing.Protocol interface
+        # Factory called when instantiating worker
+        return EchoOperation()  # User can pass anything to operation constructor
+
+class EchoOperation: # interface declaration optional; see below
```

Comment:
Hrmm, I wonder if we should have this extend an `OperationHandler` ABC that has generics for proper composability. And there can be things like `SyncOperationHandler` ABC that extends this and has an `execute` function that just returns the type.

So this class would change to something like:

```python
class EchoOperationHandler(SyncOperationHandler[EchoInput, EchoOutput]):
    def execute(self, input: EchoInput, options: StartOperationOptions) -> EchoOutput:
        return EchoOutput(input.message)
```

This makes it clearer at type checking time whether it is valid. I do think we should also have an `AsyncioOperationHandler` and `SyncAsyncioOperationHandler`. I think it is confusing to have so many runtime options, especially as extensions and helpers are created.